src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -231,101 +216,164 @@ private static bool CanOptimizeSingleSpreadAsCollectionBuilderArgument(BoundColl
             return spreadExpression is not null;
         }
 
-        private BoundExpression VisitImmutableArrayCollectionExpression(BoundCollectionExpression node, TypeWithAnnotations elementType, MethodSymbol asImmutableArray)
-        {
-            var arrayCreation = VisitArrayOrSpanCollectionExpression(
-                node,
-                CollectionExpressionTypeKind.Array,
-                ArrayTypeSymbol.CreateSZArray(_compilation.Assembly, elementType),
-                elementType);
-            // ImmutableCollectionsMarshal.AsImmutableArray(arrayCreation)
-            return _factory.StaticCall(asImmutableArray.Construct(ImmutableArray.Create(elementType)), ImmutableArray.Create(arrayCreation));
-        }
-
-        private BoundExpression VisitArrayOrSpanCollectionExpression(BoundCollectionExpression node, CollectionExpressionTypeKind collectionTypeKind, TypeSymbol collectionType, TypeWithAnnotations elementType)
+        /// <summary>
+        /// Create a collection value which may involve creating an array.
+        /// Handles types 'T[]', 'Span{T}', 'ReadOnlySpan{T}', 'ImmutableArray{T}'.
+        /// Note that Span/ROS/ImmutableArray cases, may or may not involve us actually creating an array.
+        /// Depending on the collection elements, available APIs, or other context, we may use some other strategy.
+        /// </summary>
+        private BoundExpression VisitArrayOrSpanCollectionExpression(BoundCollectionExpression node, TypeSymbol collectionType)
         {
             Debug.Assert(!_inExpressionLambda);
             Debug.Assert(_additionalLocals is { });
             Debug.Assert(node.CollectionCreation is null); // shouldn't have generated a constructor call
             Debug.Assert(node.Placeholder is null);
 
-            var syntax = node.Syntax;
-            var elements = node.Elements;
-            MethodSymbol? spanConstructor = null;
+            return collectionType switch
+            {
+                ArrayTypeSymbol arrayType
+                    => createArray(arrayType, targetsReadOnlyCollection: false),
+                TypeSymbol spanType when spanType.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_Span_T))
+                    => createSpan(isReadOnlySpan: false),
+                TypeSymbol readOnlySpanType when readOnlySpanType.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T))
+                    => createSpan(isReadOnlySpan: true),
+                TypeSymbol immutableArrayType when immutableArrayType.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_Collections_Immutable_ImmutableArray_T))
+                    => createImmutableArray(),
+                _ => throw ExceptionUtilities.UnexpectedValue(collectionType),
+            };
+
+            ArrayTypeSymbol getBackingArrayType()
```

Comment:
>getBackingArrayType [](http://example.com/codeflow?start=28&length=19)

nit: For clarity, we try to avoid captures. Please pass `collectionType` in explicitly here, and `node` to `createImmutableArray` and `createArray` below, and both to `tryCreateNonArrayBackedSpan` and `createSpan`